
# HG changeset patch
# User Karl Tomlinson <karlt+@karlt.net>
# Date 1456895040 -46800
# Node ID fe3b0ce3a5fe3d28313ecd225cd800e45c435a3b
# Parent  8b9251fa3a326b32a98dd5eb65562885b7defca2
bug 1197165 apply tooltip style class when creating window instead of when drawing r=acomminos

This is consistent with gtktooltip.c.

gtk_style_context_save() modifies the CSS node hierarchy causing this rule to
incorrectly match:

.tooltip * {
        background-color: transparent;
}

See also https://bugzilla.gnome.org/show_bug.cgi?id=761870#c2

diff --git a/widget/gtk/gtk3drawing.c b/widget/gtk/gtk3drawing.c
--- a/widget/gtk/gtk3drawing.c
+++ b/widget/gtk/gtk3drawing.c
@@ -504,16 +504,18 @@ ensure_toolbar_separator_widget()
     return MOZ_GTK_SUCCESS;
 }
 
 static gint
 ensure_tooltip_widget()
 {
     if (!gTooltipWidget) {
         gTooltipWidget = gtk_window_new(GTK_WINDOW_POPUP);
+        GtkStyleContext* style = gtk_widget_get_style_context(gTooltipWidget);
+        gtk_style_context_add_class(style, GTK_STYLE_CLASS_TOOLTIP);
         gtk_widget_realize(gTooltipWidget);
         moz_gtk_set_widget_name(gTooltipWidget);
     }
     return MOZ_GTK_SUCCESS;
 }
 
 static gint
 ensure_tab_widget()
@@ -1941,21 +1943,18 @@ moz_gtk_tooltip_paint(cairo_t *cr, GdkRe
                       GtkTextDirection direction)
 {
     GtkStyleContext* style;
 
     ensure_tooltip_widget();
     gtk_widget_set_direction(gTooltipWidget, direction);
 
     style = gtk_widget_get_style_context(gTooltipWidget);
-    gtk_style_context_save(style);
-    gtk_style_context_add_class(style, GTK_STYLE_CLASS_TOOLTIP);
     gtk_render_background(style, cr, rect->x, rect->y, rect->width, rect->height);
     gtk_render_frame(style, cr, rect->x, rect->y, rect->width, rect->height);
-    gtk_style_context_restore(style);
     return MOZ_GTK_SUCCESS;
 }
 
 static gint
 moz_gtk_resizer_paint(cairo_t *cr, GdkRectangle* rect,
                       GtkWidgetState* state,
                       GtkTextDirection direction)
 {

